### DISCUSSION TITLE
ChildDataPortal.Update yield 'Cannot wait on monitors on this runtime'

### DISCUSSION THREAD

**Discussion by clodewyks** [MarimerLLC/csla#4325] [Questions] [answered]
Hello Again!

We are experiencing a very similar issue to the previous [discussion](https://github.com/MarimerLLC/csla/discussions/4008) where it appears the sync operations which call `.Wait()` are being used instead of the async method.

Our app uses offline and online modes through `ApplicationContext.IsOffline`. When offline the Data Access Layer is storing to IndexedDB, when online it stores to SQL via the remote dataportal.

We are experiencing an issue when saving a child object from within its parents DataPortal.Insert, it appears the synchronous Update method is being called on `ChildDataPortal` instead of `UpdateAsync` only when offline, when the app is online and saving to SQL it works perfectly.

Here is the parent insert:

```
[Insert]
private async Task InsertClient([Inject] IClientDAL dataAccess, [Inject] IChildDataPortal<ClientBeneficiaries> beneficiariesDP)
{
    LoadProperty(ClientIdProperty, await dataAccess.InsertClient(LoadCientDTO()));

    if (ClientId == 0)
        throw new Exception("Data was not inserted successfully.");

     if (FieldManager.FieldExists(BeneficiariesProperty))
         await beneficiariesDP.UpdateChildAsync(ReadProperty(BeneficiariesProperty), this);
}
```

And here is the Update method on the ClientBeneficiary object:

```
[InsertChild]
async Task InsertClientBeneficiary(Client client, [Inject] IBeneficiaryDAL beneficiaryDAL)
{
    try
    {
        LoadProperty(ClientIdProperty, client.ClientId);
        SetProperty(ClientBeneficiaryIdProperty, await beneficiaryDAL.InsertClientBeneficiary(LoadDTO()));
    }
    catch (Exception ex)
    {
        throw new Exception("Data was not inserted successfully.\n" + ex.Message);
    }
}
```

When executing the InsertClientBeneficiary method we get the largely disliked `Cannot wait on monitors on this runtime` error, when looking at the stack trace it is clear that the Update method is called instead of UpdateAsync which uses `.Wait()` so the error is justified.

```
at System.Threading.Monitor.ObjWait(Int32 millisecondsTimeout, Object obj)
   at System.Threading.Monitor.Wait(Object obj, Int32 millisecondsTimeout)
   at System.Threading.ManualResetEventSlim.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.SpinThenBlockingWait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.InternalWaitCore(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.InternalWait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   **at Csla.Server.ChildDataPortal.Update(Object obj, Object[] parameters)**
   at Csla.DataPortal`1[[ClientBeneficiary, Client, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].UpdateChild(ClientBeneficiary child, Object[] parameters)
   at Csla.BusinessBindingListBase`2[[Client.Beneficiaries.ClientBeneficiaries, Client, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null],[ClientBeneficiary, Client, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].Child_Update(Object[] parameters)
   at Csla.Reflection.ServiceProviderMethodCaller.CallMethodTryAsync(Object obj, ServiceProviderMethodInfo method, Object[] parameters)'
```

The question is, why would it call the synchrounous update method?

**Comment by clodewyks** [comment]
Reverse engineering `CallMethodTryAsync` I notice the decision which method to call is based off of `IsAsyncTask` or `IsAsyncTaskObject`, I then used the same `FindDataPortalMethod` with the ClientBenficiary object, this method returns that it `IsAsyncTask` is in fact true, so we would expect the `UpdateAsync` to be called.

**Reply by rockfordlhotka** [reply]
Your child class implements two update methods with the same signature, just one is sync and the other async? Is that even allowed, since they'd only vary by return type?

**Reply by clodewyks** [reply]
Hi Rocky, there is only one [InsertChild] method in the class

**Reply by rockfordlhotka** [reply]
When you say it is calling Update instead of UpdateAsync, I am not sure I understand what methods we are talking about?

**Reply by clodewyks** [reply]
In `ServiceProviderMethodCaller.CallMethodTryAsync` there is the following code:

```
if (method.IsAsyncTask)
{
  //ClientBeneficiary is
  await ((Task)method.DynamicMethod(obj, plist)).ConfigureAwait(false);
  return null;
}
else if (method.IsAsyncTaskObject)
{
  return await ((Task<object>)method.DynamicMethod(obj, plist)).ConfigureAwait(false);
}
else
{
  var result = method.DynamicMethod(obj, plist);
  return result;
}
```

And the stacktrace shows:

`at Csla.Server.ChildDataPortal.Update(Object obj, Object[] parameters)`

When debugging I find that the ChildObjects methods is not async - but when running the same code explicitly  outside of `ServiceProviderMethodCaller` it is async. Which makes me believe you are correct that it is picking up the wrong method, but there is only the one.

**Reply by clodewyks** [reply]
I would expect that it calls `Csla.Server.ChildDataPortal.UpdateAsync(Object obj, Object[] parameters)` instead

**Reply by clodewyks** [reply]
When adding the CSLA source and adding logging to `ServiceProviderMethodCaller.CallMethodTryAsync` the method name is the expected one, and the only method with that signature, and I can confirm that is the only method with that particular name in the entire codebase

**Reply by clodewyks** [reply]
After further investigation it appears the synchronous call is happening on the Collection `ClientBeneficiaries`, not on the singular `ClientBeneficiary`. `ClientBenefciaries` does not implement any update methods and would just call the update of its children.

**Reply by rockfordlhotka** [reply]
Yes, that makes sense, as it just loops through the children, calling the child data portal to update each of them.

**Reply by clodewyks** [reply]
Agreed, so if `ServiceProviderMethodCaller.CallMethodTryAsync` logs that the Update `IsAsyncTask`, I don't understand why it would be calling `Csla.Server.ChildDataPortal.Update(Object obj, Object[] parameters)` instead of `Csla.Server.ChildDataPortal.UpdateAsync(Object obj, Object[] parameters)`

**Reply by rockfordlhotka** [reply]
Can you isolate the issue into a unit test? Or at least a simple console app that demonstrates the issue?

**Reply by clodewyks** [reply]
We will attempt to create a minimal reproduction in WASM, as this is where the Platform exception would occur.

**Reply by clodewyks** [reply, marked as answer]
@rockfordlhotka we have been unable to create a minimal repo, but we have found the issue and have a temporary workaround:

The issue is, as expected, `BusinessBindingListBase.Child_Update` which calls `ChildDataPortal.Update` which uses a `Wait()` . 

This is where it gets interesting: Why is this only an issue when using offline? When using a remote `DataPortal`, the parent call and all subsequent child calls are on the `DataPortal`, in this case a .NET Core API which allows waiting for monitors, so no issue with `.Wait()`. When running offline all `DataPortal` calls are on the client, blazor wasm, which does not support `Wait()`.

In the current CSLA source, there is a `BusinessBindingLIstBase.Child_UpdateAsync` method: [BusinessBindingListBase](https://github.com/MarimerLLC/csla/blob/30a0b0ee74f3213512c633de7b8494594e2a4ef5/Source/Csla/BusinessBindingListBase.cs#L892)

In CSLA 6.2.2 there is no such method, our temporary workaround until we can upgrade CSLA is to introduce a base class, inheriting from `BusinessBindingListBase`, implementing a `[ChildUpdate]` which is async and calls `ChildDataPortal.UpdateAsync`
